File: core/trading/xstock_strategy.py
"""
xStock / Tokenized Equity Strategy — Market hours gating and oracle validation.

xStocks trade 24/7 on Solana but their Chainlink oracle data goes stale after
4 PM ET and on weekends. Running momentum and MA strategies on stale oracle
data generates pure noise signals.

Key constraints:
    - All xStock algorithm logic MUST gate through ``is_market_hours()``
    - Oracle staleness > 5 min during market hours → degraded, widen spread
    - Oracle staleness > 30 min → halt trading entirely
    - Max backtest history: ~8 months (launched June 30, 2025)

xStock-specific slippage:
    - TSLAx, NVDAx, SPYx, AAPLx: ~1% round-trip (highest liquidity)
    - All others: 2–5% round-trip (thin pools), max 0.25% portfolio

Usage::

    from core.trading.xstock_strategy import is_market_hours, get_oracle_status

    if not is_market_hours():
        return  # Do not trade on stale oracle data

    status = get_oracle_status(last_oracle_update, current_time)
    if status.halted:
        return  # Oracle too stale
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import date, datetime, time, timezone, timedelta
from enum import Enum
from typing import Optional, Set

logger = logging.getLogger(__name__)

try:
    import pytz
    HAS_PYTZ = True
    US_EASTERN = pytz.timezone("America/New_York")
except ImportError:
    HAS_PYTZ = False
    US_EASTERN = None

# Fallback UTC offset for US Eastern (ET is UTC-5, EDT is UTC-4)
# Used only when pytz is not available
_ET_OFFSET_STANDARD = timedelta(hours=-5)
_ET_OFFSET_DST = timedelta(hours=-4)


# ---------------------------------------------------------------------------
# Market hours
# ---------------------------------------------------------------------------

MARKET_OPEN = time(9, 30)   # 9:30 AM ET
MARKET_CLOSE = time(16, 0)  # 4:00 PM ET

# U.S. market holidays for 2025-2026 (NYSE/NASDAQ)
US_MARKET_HOLIDAYS: Set[date] = {
    # 2025
    date(2025, 1, 1),    # New Year's Day
    date(2025, 1, 20),   # MLK Day
    date(2025, 2, 17),   # Presidents' Day
    date(2025, 4, 18),   # Good Friday
    date(2025, 5, 26),   # Memorial Day
    date(2025, 6, 19),   # Juneteenth
    date(2025, 7, 4),    # Independence Day
    date(2025, 9, 1),    # Labor Day
    date(2025, 11, 27),  # Thanksgiving
    date(2025, 12, 25),  # Christmas
    # 2026
    date(2026, 1, 1),    # New Year's Day
    date(2026, 1, 19),   # MLK Day
    date(2026, 2, 16),   # Presidents' Day
    date(2026, 4, 3),    # Good Friday
    date(2026, 5, 25),   # Memorial Day
    date(2026, 6, 19),   # Juneteenth
    date(2026, 7, 3),    # Independence Day (observed)
    date(2026, 9, 7),    # Labor Day
    date(2026, 11, 26),  # Thanksgiving
    date(2026, 12, 25),  # Christmas
}


def _to_eastern(dt: datetime) -> datetime:
    """Convert a datetime to US Eastern time."""
    if HAS_PYTZ:
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt.astimezone(US_EASTERN)
    else:
        # Approximate DST (second Sunday in March to first Sunday in November)
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        # Simple heuristic: March 8 – Nov 1 is roughly DST
        month = dt.month
        is_dst = 3 < month < 11 or (month == 3 and dt.day >= 8) or (month == 11 and dt.day < 1)
        offset = _ET_OFFSET_DST if is_dst else _ET_OFFSET_STANDARD
        return dt.astimezone(timezone(offset))


def is_market_hours(dt: Optional[datetime] = None) -> bool:
    """
    Return True if the given datetime falls within US equity market hours.

    Checks:
        - Weekday (Monday–Friday)
        - Between 9:30 AM and 4:00 PM Eastern
        - Not a U.S. market holiday

    If *dt* is None, uses the current UTC time.
    """
    if dt is None:
        dt = datetime.now(timezone.utc)

    et = _to_eastern(dt)

    # Weekend check
    if et.weekday() >= 5:  # Saturday=5, Sunday=6
        return False

    # Holiday check
    if et.date() in US_MARKET_HOLIDAYS:
        return False

    # Time check
    current_time = et.time()
    return MARKET_OPEN <= current_time < MARKET_CLOSE


def is_market_day(dt: Optional[datetime] = None) -> bool:
    """Return True if the date is a trading day (weekday + not holiday)."""
    if dt is None:
        dt = datetime.now(timezone.utc)
    et = _to_eastern(dt)
    if et.weekday() >= 5:
        return False
    return et.date() not in US_MARKET_HOLIDAYS


# ---------------------------------------------------------------------------
# Oracle staleness
# ---------------------------------------------------------------------------

class OracleCondition(Enum):
    """Health status of the Chainlink oracle feed."""

    FRESH = "fresh"             # < 5 minutes stale during market hours
    DEGRADED = "degraded"       # 5–30 minutes stale (widen spread)
    HALTED = "halted"           # > 30 minutes stale (do not trade)
    AFTER_HOURS = "after_hours" # Market is closed


@dataclass
class OracleStatus:
    """Oracle health assessment for an xStock."""

    condition: OracleCondition
    staleness_minutes: float
    is_market_hours: bool
    halted: bool
    spread_multiplier: float    # Multiply normal spread by this factor

    def summary(self) -> str:
        return (
            f"Oracle {self.condition.value}: "
            f"staleness={self.staleness_minutes:.1f}m, "
            f"market={'OPEN' if self.is_market_hours else 'CLOSED'}, "
            f"spread×{self.spread_multiplier:.1f}"
        )


def get_oracle_status(
    last_update: datetime,
    current_time: Optional[datetime] = None,
) -> OracleStatus:
    """
    Assess Chainlink oracle health for an xStock.

    Args:
        last_update: Timestamp of last oracle price update
        current_time: Current time (defaults to now UTC)

    Returns:
        OracleStatus with trading recommendation
    """
    now = current_time or datetime.now(timezone.utc)

    if last_update.tzinfo is None:
        last_update = last_update.replace(tzinfo=timezone.utc)
    if now.tzinfo is None:
        now = now.replace(tzinfo=timezone.utc)

    staleness_s = (now - last_update).total_seconds()
    staleness_m = staleness_s / 60

    market_open = is_market_hours(now)

    if not market_open:
        return OracleStatus(
            condition=OracleCondition.AFTER_HOURS,
            staleness_minutes=staleness_m,
            is_market_hours=False,
            halted=True,   # Do not trade after hours regardless
            spread_multiplier=5.0,
        )

    if staleness_m > 30:
        return OracleStatus(
            condition=OracleCondition.HALTED,
            staleness_minutes=staleness_m,
            is_market_hours=True,
            halted=True,
            spread_multiplier=10.0,
        )

    if staleness_m > 5:
        # Spread widens linearly: 1.0 at 5m → 3.0 at 30m
        multiplier = 1.0 + (staleness_m - 5) * (2.0 / 25)
        return OracleStatus(
            condition=OracleCondition.DEGRADED,
            staleness_minutes=staleness_m,
            is_market_hours=True,
            halted=False,
            spread_multiplier=round(multiplier, 2),
        )

    return OracleStatus(
        condition=OracleCondition.FRESH,
        staleness_minutes=staleness_m,
        is_market_hours=True,
        halted=False,
        spread_multiplier=1.0,
    )


# ---------------------------------------------------------------------------
# xStock liquidity tiers
# ---------------------------------------------------------------------------

# Tickers with highest on-chain liquidity (lower slippage)
HIGH_LIQUIDITY_XSTOCKS = frozenset({"TSLA", "NVDA", "SPY", "AAPL"})


def get_xstock_max_position_pct(underlying_ticker: str) -> float:
    """
    Maximum portfolio allocation for an xStock position.

    - High-liquidity tickers: 2% max
    - All others: 0.25% max (thin pools)
    """
    if underlying_ticker.upper() in HIGH_LIQUIDITY_XSTOCKS:
        return 0.02
    return 0.0025


def get_xstock_expected_slippage(underlying_ticker: str) -> float:
    """
    Expected round-trip slippage percentage for an xStock.

    - High-liquidity: ~1%
    - Others: ~3%
    """
    if underlying_ticker.upper() in HIGH_LIQUIDITY_XSTOCKS:
        return 0.01
    return 0.03


# ---------------------------------------------------------------------------
# Backtest constraints
# ---------------------------------------------------------------------------

# xStocks launched June 30, 2025
XSTOCK_LAUNCH_DATE = datetime(2025, 6, 30, tzinfo=timezone.utc)
XSTOCK_MAX_BACKTEST_MONTHS = 8

def validate_xstock_backtest_period(
    start_date: datetime,
    end_date: datetime,
) -> bool:
    """
    Validate that a backtest period is reasonable for xStocks.

    Returns False if start_date is before launch or span exceeds available data.
    Warns about limited statistical confidence.
    """
    if start_date < XSTOCK_LAUNCH_DATE:
        logger.warning(
            "xStock backtest start %s is before launch date %s — adjusting",
            start_date.date(), XSTOCK_LAUNCH_DATE.date(),
        )
        return False

    span_months = (end_date - start_date).days / 30
    if span_months > XSTOCK_MAX_BACKTEST_MONTHS:
        logger.warning(
            "xStock backtest span %.1f months exceeds available data (~%d months)",
            span_months, XSTOCK_MAX_BACKTEST_MONTHS,
        )

    return True

File: jarvis-sniper/src/lib/xstock-strategy.ts
/**
 * xStock Strategy - Tokenized Traditional Equities on Solana
 *
 * Handles the unique trading rules for tokenized stocks:
 * - Market hours gating (only trade during US market hours)
 * - Oracle staleness checks
 * - Higher slippage tolerance for thin pools
 *
 * @module xstock-strategy
 */

import { getUnixTime } from 'date-fns';

// US Market Hours: 9:30 AM - 4:00 PM ET, Monday-Friday
const MARKET_OPEN_HOUR = 9;
const MARKET_OPEN_MINUTE = 30;
const MARKET_CLOSE_HOUR = 16;
const MARKET_CLOSE_MINUTE = 0;

// Timezone
const US_EASTERN = 'America/New_York';

// Oracle staleness thresholds (in minutes)
const STALENESS_DEGRADED = 5;
const STALENESS_HALT = 30;

export interface XStockConfig {
  /** Symbol (e.g., 'TSLA', 'NVDA', 'AAPL') */
  symbol: string;
  /** Pool liquidity in USD */
  poolLiquidityUsd: number;
  /** Last oracle update timestamp */
  lastOracleUpdate: Date;
  /** Current SOL price for fee calculations */
  solPriceUsd: number;
}

/**
 * Check if market is open for trading
 *
 * Returns true ONLY if:
 * - Weekday (Mon-Fri)
 * - Between 9:30 AM and 4:00 PM Eastern
 * - Not a US market holiday
 */
export function isMarketHours(dt: Date): boolean {
  // Get day of week (0 = Sunday, 6 = Saturday)
  const day = dt.getDay();

  // Weekend check
  if (day === 0 || day === 6) {
    return false;
  }

  // Get hours and minutes in local time
  const hours = dt.getHours();
  const minutes = dt.getMinutes();

  // Convert to market time (simplified - assumes local is ET or handles via date-fns)
  // For simplicity, we use the input as-is assuming it's already in ET

  // Market open: 9:30
  const marketOpenMinutes = MARKET_OPEN_HOUR * 60 + MARKET_OPEN_MINUTE;
  // Market close: 16:00
  const marketCloseMinutes = MARKET_CLOSE_HOUR * 60 + MARKET_CLOSE_MINUTE;

  const currentMinutes = hours * 60 + minutes;

  // Check if within market hours
  if (currentMinutes < marketOpenMinutes || currentMinutes >= marketCloseMinutes) {
    return false;
  }

  // Check for market holidays (simplified list)
  if (isMarketHoliday(dt)) {
    return false;
  }

  return true;
}

/**
 * Check if date is a US market holiday
 */
function isMarketHoliday(dt: Date): boolean {
  const month = dt.getMonth();
  const date = dt.getDate();
  const day = dt.getDay();

  // New Year's Day (Jan 1)
  if (month === 0 && date === 1) return true;

  // Martin Luther King Jr. Day (Third Monday of January)
  if (month === 0 && day === 1 && date >= 15 && date <= 21) return true;

  // Presidents' Day (Third Monday of February)
  if (month === 1 && day === 1 && date >= 15 && date <= 21) return true;

  // Good Friday (Friday before Easter - simplified check)
  // Memorial Day (Last Monday of May)
  if (month === 4 && day === 1 && date >= 25) return true;

  // Independence Day (July 4)
  if (month === 6 && date === 4) return true;

  // Labor Day (First Monday of September)
  if (month === 8 && day === 1 && date <= 7) return true;

  // Thanksgiving (Fourth Thursday of November)
  if (month === 10 && day === 4 && date >= 22 && date <= 28) return true;

  // Christmas (December 25)
  if (month === 11 && date === 25) return true;

  return false;
}

/**
 * Get oracle staleness in minutes
 */
export function getOracleStalenessMinutes(lastUpdate: Date, currentTime: Date = new Date()): number {
  const diffMs = currentTime.getTime() - lastUpdate.getTime();
  return diffMs / (1000 * 60); // Convert to minutes
}

/**
 * Check if oracle data is fresh enough to trade
 */
export function isOracleFresh(
  lastUpdate: Date,
  currentTime: Date = new Date()
): { fresh: boolean; degraded: boolean; halted: boolean; stalenessMinutes: number } {
  const staleness = getOracleStalenessMinutes(lastUpdate, currentTime);

  if (staleness > STALENESS_HALT) {
    return { fresh: false, degraded: true, halted: true, stalenessMinutes: staleness };
  }

  if (staleness > STALENESS_DEGRADED) {
    return { fresh: false, degraded: true, halted: false, stalenessMinutes: staleness };
  }

  return { fresh: true, degraded: false, halted: false, stalenessMinutes: staleness };
}

/**
 * Determine if we can trade an xStock
 */
export function canTradeXStock(config: XStockConfig): {
  canTrade: boolean;
  reason: string;
  slippageMultiplier: number;
} {
  const now = new Date();

  // Check market hours
  if (!isMarketHours(now)) {
    return {
      canTrade: false,
      reason: 'Market closed (outside US market hours)',
      slippageMultiplier: 1,
    };
  }

  // Check oracle freshness
  const oracleStatus = isOracleFresh(config.lastOracleUpdate, now);

  if (oracleStatus.halted) {
    return {
      canTrade: false,
      reason: `Oracle stale: ${oracleStatus.stalenessMinutes.toFixed(1)} minutes`,
      slippageMultiplier: 1,
    };
  }

  // Determine slippage multiplier based on liquidity
  let slippageMultiplier = 1;
  if (config.poolLiquidityUsd < 50000) {
    slippageMultiplier = 3; // Very thin
  } else if (config.poolLiquidityUsd < 100000) {
    slippageMultiplier = 2; // Thin
  } else if (config.poolLiquidityUsd < 500000) {
    slippageMultiplier = 1.5; // Moderate
  }

  // For degraded oracle, increase slippage further
  if (oracleStatus.degraded) {
    slippageMultiplier *= 2;
  }

  return {
    canTrade: true,
    reason: oracleStatus.degraded
      ? `Oracle degraded (${oracleStatus.stalenessMinutes.toFixed(1)} min), using wider slippage`
      : 'OK',
    slippageMultiplier,
  };
}

/**
 * Get recommended slippage for xStock based on liquidity
 */
export function getXStockSlippage(
  poolLiquidityUsd: number,
  isMarketOpen: boolean
): number {
  let baseSlippage: number;

  // Base slippage by liquidity
  if (poolLiquidityUsd > 1000000) {
    baseSlippage = 0.01; // 1%
  } else if (poolLiquidityUsd > 500000) {
    baseSlippage = 0.015; // 1.5%
  } else if (poolLiquidityUsd > 100000) {
    baseSlippage = 0.02; // 2%
  } else if (poolLiquidityUsd > 50000) {
    baseSlippage = 0.03; // 3%
  } else {
    baseSlippage = 0.05; // 5% - very thin
  }

  // Increase slippage when market closed
  if (!isMarketOpen) {
    baseSlippage *= 2;
  }

  return baseSlippage;
}

/**
 * Get position size limit for xStock
 */
export function getXStockMaxPositionSize(
  poolLiquidityUsd: number,
  accountBalanceUsd: number
): number {
  // Hard cap at 0.25% of portfolio for thin xStocks
  const portfolioLimit = accountBalanceUsd * 0.0025;

  // Liquidity-based limit (can't trade more than 5% of pool)
  const liquidityLimit = poolLiquidityUsd * 0.05;

  // Return the smaller limit
  return Math.min(portfolioLimit, liquidityLimit);
}

/**
 * High liquidity xStock tickers (from Backed Finance)
 */
export const HIGH_LIQUIDITY_XSTOCKS = [
  'TSLA', // Tesla
  'NVDA', // NVIDIA
  'AAPL', // Apple
  'MSFT', // Microsoft
  'AMZN', // Amazon
  'GOOGL', // Google
  'SPY',  // S&P 500 ETF
  'QQQ',  // Nasdaq ETF
];

export function isHighLiquidityXStock(symbol: string): boolean {
  return HIGH_LIQUIDITY_XSTOCKS.includes(symbol.toUpperCase());
}
